pkg/webhook/resources/setting/validator.go

```diff
@@ -709,6 +723,159 @@ func (v *settingValidator) validateDeleteStorageNetwork(setting *v1beta1.Setting
+	prometheusVolumes := make([]string, 0)
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValuePrometheus,
+	}
+
+	pvcs, err := v.pvcCache.List(util.CattleMonitoringSystemNamespace, sets.AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pvc := range pvcs {
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+
+		prometheusVolumes = append(prometheusVolumes, pvc.Spec.VolumeName)
+	}
+
+	return prometheusVolumes, nil
+}
+
+func (v *settingValidator) getAlertManagerVolumes() ([]string, error) {
+	alertmanagerVolumes := make([]string, 0)
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueAlertManager,
+	}
+
+	pvcs, err := v.pvcCache.List(util.CattleMonitoringSystemNamespace, sets.AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pvc := range pvcs {
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+
+		alertmanagerVolumes = append(alertmanagerVolumes, pvc.Spec.VolumeName)
+	}
+
+	return alertmanagerVolumes, nil
+}
+
+func (v *settingValidator) getGrafanaVolume() (string, error) {
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueGrafana,
+	}
+
+	pvcs, err := v.pvcCache.List(util.CattleMonitoringSystemNamespace, sets.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	if len(pvcs) == 0 {
+		return "", nil
+	}
+
+	//grafana deployment only has one pvc
+	return pvcs[0].Spec.VolumeName, nil
+}
+
+func (v *settingValidator) getVMImportyControllerVolume() (string, error) {
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueImportController,
+	}
+
+	pvcs, err := v.pvcCache.List(util.HarvesterSystemNamespaceName, sets.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	if len(pvcs) == 0 {
+		return "", nil
+	}
+
+	//VM import contoller deployment only has one pvc
+	return pvcs[0].Spec.VolumeName, nil
+}
+
+func (v *settingValidator) getSystemVolumes() (map[string]struct{}, error) {
+	systemVolumes := make(map[string]struct{})
+
+	prometheusVolumes, err := v.getPrometheusVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volume := range prometheusVolumes {
+		systemVolumes[volume] = struct{}{}
+	}
+
+	alertmanagerVolumes, err := v.getAlertManagerVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volume := range alertmanagerVolumes {
+		systemVolumes[volume] = struct{}{}
+	}
+
+	grafanaVolume, err := v.getGrafanaVolume()
+	if err != nil {
+		return nil, err
+	}
+
+	if grafanaVolume != "" {
+		systemVolumes[grafanaVolume] = struct{}{}
+	}
+
+	vmImportControllerVolume, err := v.getVMImportyControllerVolume()
+	if err != nil {
+		return nil, err
+	}
+
+	if vmImportControllerVolume != "" {
+		systemVolumes[vmImportControllerVolume] = struct{}{}
+	}
+
+	return systemVolumes, nil
+}
+
+func (v *settingValidator) checkOnlineVolume() error {
+	systemVolumes, err := v.getSystemVolumes()
+	if err != nil {
+		logrus.Errorf("getSystemPVCs err: %v", err)
+		return err
+	}
+
+	for volume := range systemVolumes {
+		logrus.Infof("get getSystem volume %v", volume)
+	}
```

Comment:
why do we need to loop it just for print?

Reply:
Just for logging information, if you think it's unnecessary I can remove it.

Reply:
yes please, thanks.

Reply:
Updated per suggestion, thanks.